src/scope.rs

```diff
@@ -259,22 +252,58 @@ impl<'scope> Scope<'scope> {
             unsafe { Signal::send(&self.signal, ()) };
         }
     }
-}
 
-impl Drop for Scope<'_> {
-    fn drop(&mut self) {
-        // When the scope is dropped, block to prevent deallocation until the
-        // reference counter allows the scope to complete.
-        tracing::trace!("completing scope");
+    /// Stores a panic so that it can be propagated when the scope is complete.
+    /// If called multiple times, only the first panic is stored, and the
+    /// remainder are dropped.
+    fn store_panic(&self, err: Box<dyn Any + Send + 'static>) {
+        if self.panic.load(Ordering::Relaxed).is_null() {
+            let nil = ptr::null_mut();
+            let err_ptr = Box::into_raw(Box::new(err));
+            if self
+                .panic
+                .compare_exchange(nil, err_ptr, Ordering::Release, Ordering::Relaxed)
+                .is_ok()
+            {
+                // Ownership is now transfered into the panic field.
+            } else {
+                // Another panic raced in ahead of us, so we need to drop this one.
+                //
+                // SAFETY: This was created by `Box::into_raw` just above.
+                let _: Box<_> = unsafe { Box::from_raw(err_ptr) };
```

Comment:
Would this not resume the paused panic immediately? This needs to be leaked via `mem::forget`.

Reply:
> Does dropping the result of catch_unwind panic?

Normally no, it's just the panic payload. If you actually want to resume the panic you need to use `std::panic::resume_unwind`. 

However dropping an arbitrary value like the one held in a `Box<dyn Any>` **can panic**.

Reply:
I'll add some comments to this effect.